# Capítulo 1.

## Que es C.

**C es un lenguaje de propósito general** que ha existido casi 50 años. La versatilidad de C se debe a su diseño, al ser un lenguaje de bajo nivel, se relaciona estrechamente con la forma en que funciona la maquina sin perder simplicidad en el aprendizaje. C se ha usado para escribir casi todo para los Sistemas Operativos (OS), programas como Python, Git, base de datos Oracle entre otros.  

Nota: Entender cómo trabaja la memoria del computador es importante para la programación en el lenguaje C. 

## Hola, Mundo!

Casi siempre que aprendemos un lenguaje de programación, el programa para iniciar es un “Hola, mundo!”:
    
    #include<stdio.h>
    int main(){
        printf(“Hola, mundo!\n”); 
        return 0; 
    }

Entendamos cada línea del código: 

**#Include<stdio.h>** La función usada generalmente para definir las salidas esta en stdio.h. Para poder usar la función prinf debemos incluir el archivo que la contiene, a esto se le llama header file.  
**int main()** La función main() es el punto de entrada a un programa, los {} indican el inicio y el fin de una función (code block), lo que se coloque internamente dentro de {} determina que hará cuando se ejecute.  
**printf** Es la función usada para generar la salida.  
**\n** Genera una nueva línea.  
**;** Indica el final de la declaración.  
**return 0;** Esta declaración termina la función main(), el numero 0 indica que el programa se ha ejecutado con éxito, cualquier otro número indicara un fallo.

**Nota:** _int main(int argc, char** argv[])_ Los argumentos pasados en la línea de comandos a un programa son recibidos por la función main(). Existen dos variables predefinidas dentro del lenguaje que reciben los argumentos que se pasan al ejecutar un programa:

- **argc**, de tipo entero, contiene el número de argumentos que se han introducido, la variable argc como mínimo valdrá 1, ya que el nombre del programa se toma como primer argumento, almacenado con argv[0], que es el primer elemento de la matriz. Cada elemento del vector apunta a un argumento de la línea comandos. Todos los argumentos de la línea de comandos son cadenas.
- **argv**, de tipo array, vector (arreglo) de punteros a caracteres, argv contiene los argumentos que se han pasado desde el sistema operativo al invocar el programa.

En la siguiente línea se compilara y ejecutará el ejemplo:  

In [1]:
!gcc EC101.c -o prb && ./prb

Hola, mundo!


## Tipos de Datos 

C soporta los siguientes tipos de datos básicos: 

**int:** todos los números enteros.   
**float:** coma flotante, todos los números con parte fraccional.   
**double:** valores de coma flotante de doble precisión.   
**char:** solo un carácter.   

La cantidad de almacenamiento requerido para cada uno de estos tipos varia según la plataforma, C tiene un operador de tamaño (sizeof) que proporciona los requisitos de memoria para un tipo de dato. 

    #include <stdio.h> 
    int main(){ 
        printf(“int: %ld \n”, sizeof(int)); 
        printf(“float: %ld \n”, sizeof(float)); 
        printf(“double: %ld \n”, sizeof(double)); 
        printf(“char: %ld \n”, sizeof(char)); 
        return 0; 
    }
La salida del programa muestra el tamaño correspondiente en bytes para cada tipo de dato, la función **printf** para este programa tiene dos argumentos, el primero es la salida **string** con un formato especial (%ld), el siguiente argumento devuelve el valor de **sizeof**. En la salida final, **%ld** (para los decimales largos) es reemplazado por el valor en el segundo argumento. 
**Nota:** C no tiene un tipo booleano, los especificadores de formato también se denominan especificadores de conversión, la función prinf puede tener múltiples especificadores de formato con argumentos correspondientes para reemplazar los especificadores de conversión.

En la siguiente línea se compilara y ejecutará el ejemplo: 

In [2]:
!gcc EC102.c -o prb && ./prb

int: 4 
float: 4 
double: 8 
char: 1 


## Variables 

Una variable es el nombre para un espacio en memoria, el nombre de la variable (identifer) debe comenzar con una letra o un guion bajo y puede componerse de letras, números, etc. Existen diferentes convenciones, usualmente se usa “_” para separar palabras (soy_yo), las variables deben ser declaradas como un tipo de dato antes de ser usadas a esto se le llama declaración de asignación. Se puede declarar y asignar una variable al tiempo, además de definir variables de diferentes tipos, haciendo una operación matemática y mostrar los resultados: 

    #include <stdio.h> 
    int main(){ 
        float salario = 56.23; 
        char letra = ‘Z’; 
        int a, b; 
        a = 8; 
        b = 34; 
        int c = a+b; 
        printf(“c %d \n”, c); 
        printf(“salario %f \n”, salario); 
        printf(“letra %c \n”, letra); 
        return 0; 
    } 
Como se puede ver, se puede declarar múltiples variables en una sola línea separándola con una “,”, cabe resaltar el especificador de formato “%f” para float y “%c” para char. 

**Nota:** En el lenguaje C se hace diferencia entre mayúsculas y minúsculas, así que “soy_yo” es una variable diferente a “Soy_yo”.

En la siguiente línea se compilara y ejecutará el ejemplo: 

In [2]:
!gcc EC103.c -o prb && ./prb

c 42 
salario 56.230000 
letra Z 


## Constantes 

Una constante almacena un valor que no puede ser cambiado después de su asignación inicial, al usar nombres significativos, el código es más fácil de comprender, para diferenciar las variables de las constantes se recomienda escribirlas en mayúscula. Una forma de definir una constante es mediante la palabra const (si se asigna otro valor durante la ejecución del programa se genera un error).  

Otra manera de definir una constante es con #define, #define usa macros para definir el valor de un constante. 

    #include <stdio.h>  
    #define EULER 2.718281 
    int main(){  
        const double PI = 3.141592;  
        printf("Pi = %f \n", PI); 
        printf("Euler = %f \n", EULER); 
        return 0;  
    } 
La diferencia entre const y #define es que el primero usa memoria y el segundo no.

En la siguiente línea se compilara y ejecutará el ejemplo: 

In [3]:
!gcc EC104.c -o prb && ./prb

Pi = 3.141592 
Euler = 2.718281 


## Formato de Entrada 

La función scanf() es usada para asignar una entrada de variables. La función escanea la entrada de acuerdo con los especificadores de formato. Si la entrada no se puede convertir, la asignación no se realiza. Scanf() espera la entrada y luego realiza la asignación: 

    #include <stdio.h>
    int main(){  
        int x;
        float y; 
        char c[20];
        scanf("%d %f %s", &x, &y, c); 
        printf("%d %f %s\n", x, y, c); 
        return 0; 
    } 
Los especificadores de formato comienzan con **%**, un especificador puede incluir varias opciones: **%[*][max_field]conversion character**  
__*__ Salta el campo de entrada,  
**max_field** Da el máximo número de caracteres asignados al campo de entrada,  
**conversion character** Convierte el argumento: d decimal, c carácter, s string, f flotante, x hexadecimal. 

En la siguiente línea se compilara y ejecutará el ejemplo: 

In [9]:
## compilar en terminal
!gcc EC105.c -o prb && ./prb

/bin/sh: 12: No such file or directory


## Formato de Salida 

La función **prinf()** la usamos en nuestro primer programa, esta función puede crear una cadena de que puede incluir salidas que generan caracteres especiales y especificadores de formato que reemplazan los valores. Las sentencias de salida comienzan con un \:  
**\n** nueva línea  
**\t** pestaña horizontal  
**\\** barra invertida  
**\b** retroceso  
**\'** una frase  
**\"** comilla doble  

    #include <stdio.h> 
    int main() { 
        printf("El árbol tiene %d manzanas.\n", 22); /* salida: el árbol tiene 22 manzanas. */ 
        printf("\"Hola mundo!\"\n"); /* salida: "Hola mundo!" */ 
        printf("Color: %s, Numero: %d, flotante: %5.2f \n", "red", 42, 3.14159); /* Color: red, Numero: 42, flotante:  3.14  */ 
        printf("Pi = %3.2f \n", 3.14159); /* Pi = 3.14 */ 
        printf("Pi = %8.5f \n", 3.14159); /* Pi = 3.14159 */ 
        printf("Pi = %-8.5f \n", 3.14159); /* Pi = 3.14159 */ 
        printf("Hay %d %s en el árbol. \n", 22, "manzanas"); /* Hay 22 manzanas en el árbol. */ 
        return 0;
    } 
Los especificadores de formato comienzan con %, un especificador puede incluir varias opciones: **%[-][width].[precision]conversion character**, la opción **–** especifica la alineación izquierda de los datos, **width** proporciona el número mínimo de caracteres para los datos, **precision** especifica la cantidad de decimales después de la coma(punto), si se usa s, la precisión determina el número de caracteres a imprimir. 

En la siguiente línea se compilara y ejecutará el ejemplo: 

In [6]:
!gcc EC106.c -o prb && ./prb

El árbol tiene 22 manzanas.
"Hola mundo!"
Color: red, Numero: 42, flotante:  3.14 
Pi = 3.14 
Pi =  3.14159 
Pi = 3.14159  
Hay 22 manzanas en el árbol. 


## Comentarios 

Los comentarios son información explicativa para beneficiar al lector del código. El compilador ignora los comentarios, por lo que no tienen ningún efecto en un programa. Un comentario comienza con /* y termina con */ y puede estar en cualquier parte del código. 

    #include <stdio.h> 
    /* Un simple programa en C
     *  Version 1.0 
     */ 
    int main() { 
        /* Dar salida a una cadena */ 
        printf("Hello World!"); 
        return 0; 
    } 
C ++ introdujo un comentario de barra doble // como una forma de comentar líneas simples. Algunos compiladores de C también admiten este estilo de comentario. 

En la siguiente línea se compilara y ejecutará el ejemplo:

In [7]:
!gcc EC107.c -o prb && ./prb

Hello World!

## Operadores aritméticos 

C admite operadores aritméticos + (suma), - (sustracción), * (multiplicación), / (división) y % (división de módulo). Los operadores a menudo se usan para formar una expresión numérica como 10 + 5, las expresiones numéricas se usan a menudo en las declaraciones de asignación. 

    #include <stdio.h> 
    /// Programa
    int main() { 
        int length = 10, width = 5, area; 
        area = length * width; 
        printf("%d \n", area);  /* 50 */ 
        //---------------------------------- 
        int i1 = 10, i2 = 3, cociente, resto; 
        float f1 = 4.2, f2 = 2.5, result; 
        quotient = i1 / i2;  // 3 
        remainder = i1 % i2;  // 1 
        result = f1 / f2;  // 1.68 
        printf("%d %d %f \n", cociente, resto, result); 
        //---------------------------------- 
        return 0; 
    } 

C tiene dos operadores de división: **/** y **%**. **/** funciona de manera diferente según los tipos de datos. Cuando ambos operandos son tipos de datos int, la división de enteros, también llamada división truncada, elimina cualquier resto para dar como resultado un entero. Cuando uno o ambos operandos son números reales (flotante o doble), el resultado es un número real. **%** devuelve solo el resto de la división de enteros. Es útil para muchos algoritmos, incluida la recuperación de dígitos de un número. El módulo no se puede realizar en flotantes o dobles. 

C evalúa una expresión numérica basada en la precedencia del operador. Los **+** y **-** son iguales en precedencia, al igual que __*__, **/** y **%**. Pero *, / y % se realizan primero en orden de izquierda a derecha y luego + y -, también en orden de izquierda a derecha. 

Puede cambiar el orden de las operaciones utilizando paréntesis **()** para indicar qué operaciones se deben realizar primero. Ejp: 5 + 3 * 2 = 11 o  (5 + 3) * 2 = 16. 

En la siguiente línea se compilara y ejecutará el ejemplo:

In [8]:
!gcc EC108.c -o prb && ./prb

50 
3 1 1.680000 


## Conversión Explicita

En C, cuando una expresión numérica contiene operandos de diferentes tipos de datos, se convierten automáticamente según sea necesario en un proceso llamado conversión de tipo. Ejp: en una operación que involucra tanto flotantes como enteros, el compilador convertirá los valores enteros en valores flotantes. Cuando desee forzar el resultado de una expresión a un tipo diferente, puede realizar una conversión explícita. 

    #include <stdio.h> 
    int main() { 
        //-----------conversión automática 
        float price = 6.50, new_price; 
        int increase = 2; 
        new_price = price + increase; 
        printf("El Nuevo precio es %4.2f \n", new_price); 
        //-----------conversión explicita 
        float average; 
        int total = 23, count = 4; 
        average = (float) total / count; 
        printf("El promedio es %4.2f", average); 
        //----------------------------------------- 
        return 0; 
    } 
Incluso cuando el compilador puede realizar una conversión de tipo automática, la conversión de tipo explícita se considera una buena práctica de programación. 

En la siguiente línea se compilara y ejecutará el ejemplo:

In [9]:
!gcc EC109.c -o prb && ./prb

El Nuevo precio es 8.50 
El promedio es 5.75

## Operadores de Asignación 

La asignación evalúa primero la expresión en el lado derecho del __=__ y luego asigna ese valor a la variable en el lado izquierdo. Esto hace posible usar la misma variable en ambos lados de una declaración de asignación, para acortar este tipo de declaración de asignación, C ofrece el operador de asignación +=. Muchos operadores C tienen un operador de asignación correspondiente. 

    #include <stdio.h> 
    int main() { 
        int x = 2; 
        x = x + 1;  /* x es ahora 3 */ 
        //x += 1;  // 3 
        //x -= 1;  // 1 
        //x *= 3;  // 6 
        //x /= 2;  // 1 
        //x %= 2;  // 0 
        //x += 3 * 2;  // 8 
        printf("%d", x); 
        return 0; 
    } 
En la siguiente línea se compilara y ejecutará el ejemplo:

In [10]:
!gcc EC110.c -o prb && ./prb

3

## Incremento y Disminución 

Agregar 1 a una variable se puede hacer con el operador de incremento ++. De manera similar, el operador de decremento -- se usa para restar 1 de una variable. Los operadores de incremento y decremento pueden usarse ya sea prefix (antes del nombre de la variable) o postfix (después del nombre de la variable). La forma en que usa el operador puede ser importante en una declaración de asignación. 

    #include <stdio.h> 
    int main() { 
        int x, y, z; 
        //------------------------------------------------- 
        z = 3; 
        x = z--;  /* asigna 3 a x, y reduce z a 2 */ 
        printf("x = %d \n", x); 
        printf("z = %d \n", z);
        //-------------------------------------------------     
        y = 3; 
        x = ++y;  /* incrementa y a 4, asigna 4 a x */ 
        printf("x = %d \n", x); 
        printf("y = %d \n", y); 
        //------------------------------------------------- 
        return 0; 
    }
En la siguiente línea se compilara y ejecutará el ejemplo:

In [13]:
!gcc EC111.c -o prb && ./prb

x = 3 
z = 2 
x = 4 
y = 4 
